src/pybind/mgr/devicehealth/module.py

```diff
@@ -304,6 +323,149 @@ def show_device_metrics(self, devid, sample):
+                device_locations = map(lambda x: x['host'] + ':' + x['dev'],
+                                       dev['location'])
+                health_warnings['DEVICE_HEALTH'].append(
+                    '%s (%s); daemons %s; life expectancy between %s and %s'
+                    % (dev['devid'],
+                       ','.join(device_locations),
+                       ','.join(dev.get('daemons', ['none'])),
+                       dev['life_expectancy_min'],
+                       dev.get('life_expectancy_max', 'unknown')))
+                # TODO: by default, dev['life_expectancy_max'] == '0.000000',
+                # so dev.get('life_expectancy_max', 'unknown')
+                # above should be altered.
+
+        # OSD might be marked 'out' (which means it has no
+        # data), however PGs are still attached to it.
+        for _id in osds_out.iterkeys():
+            num_pgs = self.get_osd_num_pgs(_id)
+            if num_pgs > 0:
+                health_warnings['DEVICE_HEALTH_IN_USE'].append(
+                    'osd.%s is marked out '
+                    'but still has %s PG(s)' %
+                    (_id, num_pgs))
+        if osds_in:
+            self.log.debug('osds_in %s' % osds_in)
+            # calculate target in ratio
+            num_osds = len(osdmap['osds'])
+            num_in = len([x for x in osdmap['osds'] if x['in']])
+            num_bad = len(osds_in)
+            # sort with next-to-fail first
+            bad_osds = sorted(osds_in.items(), key=operator.itemgetter(1))
+            did = 0
+            to_mark_out = []
+            for osd_id, when in bad_osds:
+                ratio = float(num_in - did - 1) / float(num_osds)
+                if ratio < min_in_ratio:
+                    final_ratio = float(num_in - num_bad) / float(num_osds)
+                    checks['DEVICE_HEALTH_TOOMANY'] = {
+                        'severity': 'warning',
+                        'summary': health_messages['DEVICE_HEALTH_TOOMANY'],
+                        'detail': [
+                            '%d OSDs with failing device(s) would bring "in" ratio to %f < mon_osd_min_in_ratio %f' % (num_bad - did, final_ratio, min_in_ratio)
+                        ]
+                    }
+                    break;
+                to_mark_out.append(osd_id)
+                did += 1
+            if to_mark_out:
+                self.mark_out_etc(to_mark_out)
+        for warning, ls in health_warnings.iteritems():
+            n = len(ls)
+            if n:
+                checks[warning] = {
+                    'severity': 'warning',
+                    'summary': health_messages[warning] % n,
+                    'detail': ls,
+                }
+        self.set_health_checks(checks)
+        return (0,"","")
+
+    def is_osd_in(self, osdmap, osd_id):
+        for osd in osdmap['osds']:
+            if str(osd_id) == str(osd['osd']):
+                return osd['in']
+        # return False
```

Comment:
I'd uncomment this, it's cleaner to return False than to return None on an `is_` type function.

I'd also do a bool(osd['in']) on the main return path, so that this function isn't slightly confusingly returning a number.